engine/function/src/templates/Numeric.ftl

```diff
@@ -1820,6 +1841,32 @@ public class Numeric {
         return (${pt.primitive})(lowerBin((${pt.primitive})(value-offset),interval) + offset);
     }
 
+    <#if pt.valueType.isFloat >
+
+    /**
+     * Returns the upper bound of the bin containing the value.
+     *
+     * @param value value.
+     * @param interval bin width.
+     * @return upper bound of the bin containing the value.
+     */
+    public static ${pt.primitive} upperBin(${pt.primitive} value, ${pt.primitive} interval) {
```

Comment:
I'm not exactly sure what the definition of "upperBin" should be, but I felt a little "surprised" by the assymetry between lowerBin and upperBin:

| v | lowerBin(v,3) | upperBin(v,3) |
|--|------------|-------------|
| 0 | 0 | 0 |
| 1 | 0 | 3 |
| 2 | 0 | 3 | 
| 3 | 3 | 3 | 
| 4 | 3 | 6 |
| 5 | 3 | 6 |
| 6 | 6 | 6 |
| 7 | 6 | 9 |
| 8 | 6 | 9 |
| 9 | 9 | 9 |
| 10 |  9 |  12 |
| 11 | 9 |  12 |

I guess you're going for an analogy to floor() and ceil() so maybe it's fine.

In any case, I think maybe it might help if there were a comment with an algebraic relation like

```
lowerBin: if value >= k * interval and value < (k + 1) * interval for some integer k, return k * interval
```

(Also I'm kind of punting on what happens if you pass in negative numbers... I'm not sure that's legal... do we want to check for negative numbers and throw?)

and then the definition for upperBin is something like

```
upperBin: if value > k * interval and value <= (k + 1) * interval for some integer k, return (k + 1) * interval
```
Which I *think* matches the code you've written. Thinking about the sort-of-special-case zero, that works too, if we allow k to be -1.

So to summarize I guess I'm requesting that we have a crisp algebraic formula like the above in the comments. Or some other way of expressing it if you prefer.

Reply:
That code was a mess.  I _think_ it is now resolved.